### ISSUE TITLE
Multidimensional arrays with different start indices

### ISSUE THREAD

**Issue by marshallward** [marshallward/f90nml#93]
If a multidimensional array uses (presumed) different start indices in different dimensions, then there may be errors.  Example:
```
&a
    alpha(1,3) = 4
    alpha(0:4,6) = 0.0, 0.0, 0.012, 0.012, 0.012
/
```
will produce the following incorrect namelist
```
&a
    alpha(0:4,3) = , , , , ,
    alpha(0:4,4) = 4, , , , ,
    alpha(0:4,5) = , , , , ,
    alpha(0:4,6) = 0.0, 0.0, 0.012, 0.012, 0.012
/
```
That is, the value in `alpha(1,3)` somehow gets stored to `alpha(0,4)`.

The bug appears to be at the `Parser` level, and not the `Namelist` output.

This particular case can be averted by using `p.global_start_index = 0` for a `Parser` `p`, but I consider this to be a very serious bug and should be resolved ASAP.

**Comment by marshallward** [comment]
My guess here is that writing `alpha(1,3)` is setting the `start_index` to `(1, 3)`, and that the subsequent write to `alpha(0:4, 6)` is updating the `start_index` to `(0, 3)` but is not moving or replacing the original values.

**Comment by marshallward** [comment]
Confirmed that swapping the lines:
```
&a
    alpha(0:4,6) = 0.0, 0.0, 0.012, 0.012, 0.012
    alpha(1,3) = 4
/
```
gives the correct output:
```
&a
    alpha(0:1,3) = , 4
    alpha(0,4) = ,
    alpha(0,5) = ,
    alpha(0:4,6) = 0.0, 0.0, 0.012, 0.012, 0.012
/
```
(albeit in our wonky padded multidimensional format)

It seems our tests are not actually considering this possibility, and we will need to account for existing values when changing `start_index`.

**Comment by marshallward** [comment]
There seem to be many issues with multidimensional indices with decreasing start index, but this is one cruicial problem:
```
                 # Resize vector based on starting index
                 for i_p, i_v in zip(p_idx, v_idx.first):
                     if i_p is not None and i_v is not None and i_v < i_p:
                         pad = [None for _ in range(i_p - i_v)]
                         parent[v_name] = pad + parent[v_name]
```
This loop is correctly detecting the padding across each dimension, but then incorrectly applies all of the padding onto the first rank (`parent[v_name]`) and is not penetrating across the subaxes, e.g. `parent[v_name][i_p1][i_p2]...`.

**Comment by marshallward** [comment]
Commit de1f2ad69e773d29085305c9dd2da5f02e332bf9 seems to solve the issue mentioned above with respect to padding, but does not solve the overall problem observed in the original case.  The hunt continues...

**Comment by marshallward** [comment]
This is a simpler case which highlights the problem.

This namelist:
```
&a
   m(1,1) = 1
   m(3,2) = 2
/
```
gives the correct output:
```
&a
    m(1:3,1) = 1, , ,
    m(1:3,2) = , , 3
/
```

But if I reverse the records:
```
&a
    m(3,2) = 3
    m(1,1) = 1
/
```
then I get an incorrectly sized array:
```
&a
    m(1,1) = 1
    m(1,2) = ,
    m(1:2,3) = , 3
/
```
This is likely a Fortran indexing order error somewhere.

**Comment by marshallward** [comment]
This appears to be fixed in 8cca2def863fdba156d8624da3b55a6b79befba2, and all tests are passing (including new out-of-order multidimensional tests) so closing this issue.
